feat: Add `--autoplan-file-list` server flag.

Omicron7 opened the pull request:
This adds a new server flag that allows modifying the list of files that trigger default project planning.

We use terraform/atlantis to build AWS AMI's using packer and ansible. We wanted changes to the packer (`*.pkr.hcl`) and ansible (`*.yml`) files to trigger an autoplan when modified in a commit. It looked like we could do this with an `atlantis.yml` file (Workflow), but we have a large repo with 100's of small projects and rely on the default atlantis workflow. Defining `atlantis.yml` would require us to define every existing project, and all new projects in that workflow.

Instead, I added a server flag/ENV that would allow modifying/overriding the list of file patterns that trigger an autoplan.

`--autoplan-file-list` uses the same dockeringore syntax as Custom workflows autoplan `when_modified`. This allows for overriding the default list, adding more file types, and file exclusions.

Examples:
* Default: `--autoplan-file-list="**/*.tf,**/*.tfvars,**/*.tfvars.json"`
* Add packer and YML files: `--autoplan-file-list="**/*.tf,**/*.tfvars,**/*.tfvars.json,**/*.pkr.hcl,**/*.yml"`
* ENV var: `ATLANTIS_AUTOPLAN_FILE_LIST="**/*.tf,**/*.hcl"`

nishkrishnan commented on server/events/project_finder.go:
```diff
@@ -145,18 +148,34 @@ func (p *DefaultProjectFinder) DetermineProjectsViaConfig(log *logging.SimpleLog
+			log.Debug("filter err for file %q: %s", fileName, err)
+			continue
+		}
+		if match || filepath.Base(fileName) == "terragrunt.hcl" {
```
since the syntax changed to dockerignore syntax can we just add `**/terragrunt.hcl` to the fileList? I might be wrong, but wouldn't that look for the exact file as well?

Omicron7 replied:
You are correct. I'll remove the or clause here and add terragrunt.hcl to the defaults.

nishkrishnan commented on server/events/project_finder.go:
```diff
@@ -48,12 +47,16 @@ var ignoredFilenameFragments = []string{"terraform.tfstate", "terraform.tfstate.
-	modifiedTerraformFiles := p.filterToTerraform(modifiedFiles)
+	modifiedTerraformFiles, err := p.filterToFileList(log, modifiedFiles, autoplanFileList)
+	if err != nil {
+		return nil, errors.Wrapf(err, "filtering modified files by autoplanFileList: %s", autoplanFileList)
```
seems unnecessary to wrap this error again, how about logging it instead?

Omicron7 replied:
I wasn't sure what to do here. Contributing docs suggested always wrapping errors. When I move the default value out of the function. I'm going to add the syntax error checking to the attribute validation step. That would mean we don't need to worry about errors here anymore, so this will go away.

nishkrishnan reviewed (commented):
Nice work I'm in favor of the change just a couple questions/comments.

Omicron7 commented on server/events/project_finder.go:
```diff
@@ -145,18 +148,34 @@ func (p *DefaultProjectFinder) DetermineProjectsViaConfig(log *logging.SimpleLog
+func (p *DefaultProjectFinder) filterToFileList(log *logging.SimpleLogger, files []string, fileList string) ([]string, error) {
 	var filtered []string
-	fileNameRe, _ := regexp.Compile(`^.*(\.tf|\.tfvars|\.tfvars.json)$`)
+	if fileList == "" {
+		fileList = "**/*.tf,**/*.tfvars,**/*.tfvars.json"
+	}
```
Thanks for the feedback. I tried doing that originally, but found myself repeating the default string a bunch of times in test files, which is probably OK. I did try to use the default variable from cmd/server.tf, but got a cyclic import issue. I'm fairly new to golang, and wasn't sure how to handle a default value that is shared between tests and code without repeating it a bunch of times.

I do like this idea, so I'll work on moving it to the default.

Omicron7 commented:
@nishkrishnan I've updated the PR to set the default value in cmd/server.go and added `**/terragrunt.hcl` to the list. This also adds pattern validation at startup which removes the need for error checking/catching in the project finder.

Omicron7 commented:
`make test` passes for me locally, not sure why circleci failed on the `invalid pattern` test.

Omicron7 commented:
@nishkrishnan I rebased this on latest master and fixed logging issue. Any help tracking down the failing test would be appreciated. `make test` succeeds locally for me. I'm not knowledgeable enough with go and circleci to figure out whats going on. I'm guessing a module version difference.
